### ISSUE TITLE
Provide React Hooks

### ISSUE THREAD

**Issue by vincentjames501** [redux-form/redux-form#4317]
Has there been any thoughts on React Hooks adoption vs using a HOC? I know things are early and even Redux is still working to get a full grasp on how to do things in the React Hooks world ([here](https://github.com/facebook/react/issues/14110), and [here](https://github.com/reduxjs/react-redux/issues/1063)), but just thought I'd throw this out there for discussion and thoughts of what it _might_ look like once Redux gets proper hook support.

```js
function MyForm() {
  const { handleSubmit, context, ...otherProps } = useReduxForm({form: 'myForm', ...otherConfigProps});
  return (
    <Form context={context} onSubmit={() => console.log('submitted!')} >
      <Field name="myInput" component="input" type="text" />
    </Form>
  );
}
```

**Comment by gabrielbs** [comment]
Has this been answered somewhere?

**Comment by aislanmaia** [comment]
This doesn't seems getting any attention at all in this library. But Hooks could be the dream for this because it will remove all the boilerplate and complexity regarding setup the form just to work.

**Comment by fzaninotto** [comment]
[React-Redux v7.1](https://github.com/reduxjs/react-redux/releases/tag/v7.1.0-alpha.1) (currently in alpha) offers hook alternatives to `connect` (namely `useDispatch` and `useSelector`). Everything is ready to support hooks in redux-form. 

Is there any interest for this feature in the core team?

**Comment by JerryGreen** [comment]
I think there's no reason to use `redux-form` since this:

[README#️attention](https://github.com/erikras/redux-form/commit/220134950b9a383554021838f6fdbf74082f6828#diff-04c6e90faac2675aa89e2176d2eec7d8R24)

And this:

https://github.com/final-form/react-final-form
https://github.com/final-form/react-final-form-hooks

And `redux`, too. Use [`useContext`](https://reactjs.org/docs/hooks-reference.html#usecontext) instead (with other hooks like [`useReducer`](https://reactjs.org/docs/hooks-reference.html#usereducer))

**Comment by fzaninotto** [comment]
About using react-final-form vs redux-form, this has already been discussed earlier in this issue. Di you have any objective arguments or experience to help us choose between react-final-form and formik?

About redux vs useContext, you probably don't know all that react-redux does. Check [the latest react-redux implementation](https://github.com/reduxjs/react-redux/blob/master/src/components/connectAdvanced.js); it uses hooks, and it uses MUCH MORE than just useContext. We won't remove Redux from the mix in react-admin because we don't want to reinvent that logic.

**Comment by iamandrewluca** [comment]
Providing react hooks, we will need to bump react version to 16.8, and this is a breaking change from the library point of view. But as we know react is back compatible and users of `redux-form` can just upgrade to 16.8 without problems.
@erikras @renatoagds

**Comment by pokonski** [comment]
@JerryGreen sure, let me rewrite a hundred forms to another library just to have a hook ;)

**Comment by iamandrewluca** [comment]
@pokonski adding hooks should not affect curent components behaviour.
We can just create the hooks, and leave the Components as is.
Or we can create hooks remove duplicate logic, and Components will use hooks under the hood.

**Comment by jedwards1211** [comment]
@iamandrewluca you could avoid the breaking change in required react version if existing components don't use hooks under the hood and users have to import the hooks directly from a specific submodule/subfolder like `import {useFormValue} from 'redux-form/hooks'`.  That module could throw an error when first loaded if the version of React doesn't support hooks.

**Comment by iamandrewluca** [comment]
@jedwards1211 yes this is possible, but then we will have to maintain 2 codebases, one with hooks, and one without.

**Comment by jedwards1211** [comment]
I think the people who came up with semver would be very angry about that idea.  If you publish a minor update any developers with
```js
"react": "16.7.0",
"redux-form": "^8.0.0",
```
will get a nasty unexpected surprise.  Please don't contribute to developer suffering.  Just do the right thing and release it as a major version bump.

**Comment by jedwards1211** [comment]
Example of responsible semver usage:
https://github.com/mui-org/material-ui/blob/master/CHANGELOG.md#material-uicorev400-alpha0
### `@material-ui/core@v4.0.0-alpha.0`
##### Breaking changes
* [core] Increase React peer dependency to v16.8.0 (#14432) @oliviertassinari

**Comment by iamandrewluca** [comment]
@jedwards1211 https://semver.org/#what-should-i-do-if-i-update-my-own-dependencies-without-changing-the-public-api

**Comment by jedwards1211** [comment]
@iamandrewluca that section addresses dependencies used internally, not peer dependencies.  The bottom line of semver is, if any project would break if the new release is dropped in, it should be released as a new major version.

**Comment by jedwards1211** [comment]
Apollo v3.0.0 release from the changelog https://github.com/apollographql/react-apollo/blob/master/Changelog.md#breaking-changes
### Breaking changes
* The minimum supported React version is now 16.8.

**Comment by jedwards1211** [comment]
`react-jss` v10.0.0-alpha.1 https://github.com/cssinjs/jss/blob/master/changelog.md#breaking-changes-3
### Breaking changes
* [react-jss] Drop support for older React versions, require v16.3 or higher (#868, #851)

**Comment by jedwards1211** [comment]
`react-redux` 7.0.0-beta.0 https://github.com/reduxjs/react-redux/releases/tag/v7.0.0-beta.0

"The major change for this release is that connect is now implemented using Hooks internally. Because of this, we now require a minimum React version of 16.8.4 or higher."

**Comment by iamandrewluca** [comment]
@jedwards1211 ok, I got it :slightly_smiling_face:

**Comment by pokonski** [comment]
@iamandrewluca I meant in my reply that his advice to rewrite to a completely different form library is reckless suggestion. That would be weeks of work for a small convience.

**Comment by Firanolfind** [comment]
any updates?

**Comment by iamandrewluca** [comment]
At this moment I think no one is interested in allocating time for `redux-form` maintance, and feature implementations.
We have a v9 roadmap, but for now is on hold, because we don't have enough resources.

https://github.com/redux-form/redux-form/issues/4467

**Comment by shorif2000** [comment]
> You can hack it in yourself at least:
> 
> ```js
> /**
>  * @flow
>  */
> import { useContext } from 'react'
> import { ReduxFormContext } from 'redux-form/es/ReduxFormContext'
> import { type Context } from 'redux-form/es/types'
> 
> export default function useReduxForm(): Context {
>   return useContext(ReduxFormContext)
> }
> ```
> 
> ```js
> /**
>  * @flow
>  */
> 
> import useReduxForm from './useReduxForm'
> import { useSelector } from 'react-redux'
> import getFormValues from 'redux-form/es/getFormValues'
> 
> export default function useFormValue(name: string): any {
>   const { form } = useReduxForm()
>   return useSelector(state => getFormValues(form)(state)?.[name])
> }
> ```

Can you provide full page example with a form?
